pymeasure/instruments/keithley/keithley2700.py

```diff
@@ -0,0 +1,362 @@
+        rows = np.array(rows)
+        columns = np.array(columns)
+
+        channels = (rows - 1) * 8 + columns
+
+        if slot is not None:
+            channels += 100 * slot
+
+        return channels
+
+    # system, some taken from Keithley 2400
+    def beep(self, frequency, duration):
+        """ Sounds a system beep.
+
+        :param frequency: A frequency in Hz between 65 Hz and 2 MHz
+        :param duration: A time in seconds between 0 and 7.9 seconds
+        """
+        self.write(":SYST:BEEP %g, %g" % (frequency, duration))
+
+    def triad(self, base_frequency, duration):
+        """ Sounds a musical triad using the system beep.
+
+        :param base_frequency: A frequency in Hz between 65 Hz and 1.3 MHz
+        :param duration: A time in seconds between 0 and 7.9 seconds
+        """
+        self.beep(base_frequency, duration)
+        time.sleep(duration)
+        self.beep(base_frequency * 5.0 / 4.0, duration)
+        time.sleep(duration)
+        self.beep(base_frequency * 6.0 / 4.0, duration)
+
+    @property
+    def error(self):
+        """ Returns a tuple of an error code and message from a
+        single error. """
+        err = self.values(":system:error?")
+        if len(err) < 2:
+            err = self.read()  # Try reading again
+        code = err[0]
+        message = err[1].replace('"', '')
+        return (code, message)
+
+    def check_errors(self):
+        """ Logs any system errors reported by the instrument.
+        """
+        code, message = self.error
+        while code != 0:
+            t = time.time()
+            log.info("Keithley 2700 reported error: %d, %s" % (code, message))
+            print(code, message)
+            code, message = self.error
+            if (time.time() - t) > 10:
+                log.warning("Timed out for Keithley 2700 error retrieval.")
+
+    def reset(self):
+        """ Resets the instrument and clears the queue.  """
+        self.write("status:queue:clear;*RST;:stat:pres;:*CLS;")
+
+    options = Instrument.measurement(
+        "*OPT?",
+        """Property that lists the installed cards in the Keithley 2700.
+        Returns a dict with the integer card numbers on the position.""",
+        cast=False
+    )
+
+    ###########
+    # DISPLAY #
+    ###########
+
+    text_enabled = Instrument.control(
+        "DISP:TEXT:STAT?", "DISP:TEXT:STAT %d",
+        """ A boolean property that controls whether a text message can be
+        shown on the display of the Keithley 2700.
+        """,
+        values={True: 1, False: 0},
+        map_values=True,
+    )
+    display_text = Instrument.control(
+        "DISP:TEXT:DATA?", "DISP:TEXT:DATA '%s'",
+        """ A string property that controls the text shown on the display of
+        the Keithley 2700. Text can be up to 12 ASCII characters and must be
+        enabled to show.
+        """,
+        validator=text_length_validator,
+        values=12,
```

Comment:
This is an interesting use of the validator pattern. As more cases like this occur, it will be worth revisiting the syntax to see if there are better ways to generalize. In the meantime, I think this is still clear.